src/librustc_trans/abi.rs

```diff
@@ -1055,6 +1055,27 @@ impl<'a, 'tcx> FnType<'tcx> {
             PassMode::Indirect(ref attrs) => apply(attrs),
             _ => {}
         }
+        if let layout::Abi::Scalar(ref scalar) = self.ret.layout.abi {
```

Comment:
This duplicates https://github.com/rust-lang/rust/blob/master/src/librustc_trans/mir/place.rs#L93:L116 right?

Reply:
Partially yes, there are two differences:

* no `nonnull` metadata is ever emitted for calls, because that's illegal;
* boolean scalars don't get `range` metadata because they end up as `i1` with a<br>`!{i1 false, i1, false}` range otherwise.

Reply:
I guess I could make the range extraction and masking its own method on `Scalar`, would you prefer that?

Reply:
Good point about `nonnull`, but the calculation of the range (and the relevant comments) should still be de-duplicated.

The i1 part confuses me though. Surely the same problem applies to range metadata on loads? If it is a problem at all, that is: the `max_next & mask != min & mask` guard should rule that out already, I think? Did something break without the extra check for bools?

Reply:
About the `i1` part: https://botbot.me/mozilla/rustc/2018-04-22/?msg=99259759&page=3

I'll add some comments.

Reply:
> I guess I could make the range extraction and masking its own method on Scalar, would you prefer that?

I still don't know the layout code well enough for my opinion on this to carry much weight, but this is really an LLVM concept, so maybe it should be in rustc_trans. No real clue where, though.

Note that LLVM ranges with `max_next & mask == min & mask` are inherently invalid, so the helper function could return `Option<Range<_>>` to avoid duplicating that check as well.